src/coreclr/gc/gc.cpp

```diff
@@ -35961,6 +35968,13 @@ void gc_heap::background_promote_callback (Object** ppObject, ScanContext* sc,
         return;
     }
 
+#if defined(FEATURE_CONSERVATIVE_GC) && defined(USE_REGIONS)
+    if (!(is_in_bookkeeping_range (o)))
```

Comment:
This is a check that does not required lookup of the heap. Should be it moved up before we lookup the heap, and maybe even replace the null check?

Something like:
```
 #if defined(FEATURE_CONSERVATIVE_GC) && defined(USE_REGIONS)
    if (!(is_in_bookkeeping_range (o)))
    {
        return false;
    }
    // null should be never included in the bookkeeping range 
    assert(o != nullptr);
#else
    if (o == nullptr)
    {
        return;
    }
#endif
```

Reply:
> and maybe even replace the null check

Probably not. null is very common and so it makes sense to have dedicated early out for it.

The part of my comment about moving it up still applies.